news/_posts/2020-11-30-from-legacy-to-future-architecture-connecting-the-dots.md

```diff
@@ -0,0 +1,308 @@
+
+Now, let's explore those "chunks" individually.
+
+
+
+## Reduce entropy to increase reliability
+
+A stable system relies on pre-established business rules being respected at all times. If a system's business logic requires all products to have a name and a non-negative price, then it is _expected_ that if a product exists, it will adhere to these rules. Unfulfilled expectations produce inconsistencies, which result in errors.
+
+PrestaShop is a very complex system. As such, it needs to enforce a very high number of business rules corresponding to a wide range of use cases. One of the biggest problems of the current architecture is that extensions can tap anywhere into the Core: since the system is very complex, extensions hooking, overriding or just _using_ Core classes can lead to problems if some business rule is overlooked by a developer who was most likely not even aware of the existence of such rule.
+
+For example, when an order's carrier is updated, an email should be sent to the customer informing them that the order is on its way. This is done automatically when the action is performed using the appropriate service. However, if the order's carrier is modified manually through ObjectModel, for example by a module developer not aware of this rule, then the required email is not sent.
+
+This problem of thick service layers manipulating thin models devoid of domain logic was identified over 15 years ago by Martin Fowler, who described it the [Anemic Domain Model][anemic-domain-model] anti-pattern. His answer was simple: business logic should be put in the same place as data is – [entity objects][ddd-entities].
+
+This means that **there should be one single, obvious way to perform any given system action**. If there is only one possible way to update an order's carrier, and if this is done in an intuitive way through the Order entity, then we can ensure that all business rules are being enforced regardless of where this task is initiated.
+
+Here is how you would update an order's shipping details:
+
+```php
+$someCustomer // this is a Customer
+  ->getOrderById($orderId) // this returns an Order
+	->updateShippingDetails($newCarrier, $trackingNumber);
+```
+
+Of course, this is easier said that done: getting to this kind of workflow in PrestaShop would require a complete rewrite of the Core, and as we said earlier, we don't want to do that. Which brings us to the next issue...
+
+### Extensions need a reliable interface to the Core
```

Comment:
2-level title or 3-level ?

Reply:
I thought about it, I think it's a continuation of the previous 2nd level section

- Entropy must be reduced by changing the way code is consumed
- We can't do that because it's a breaking change

Technically "Isolate the Core behind the API" should also be a 3rd level title, but I though it deserved a section of its own. We can think of the first two 2nd level titles as introductory.